src/storage/hummock_sdk/src/version.rs

```diff
@@ -12,27 +12,336 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use std::collections::{HashMap, HashSet};
+use std::collections::{BTreeMap, HashMap, HashSet};
 use std::mem::size_of;
 use std::sync::Arc;
 
-use prost::Message;
+use bytes::Bytes;
+use itertools::Itertools;
 use risingwave_common::catalog::TableId;
+use risingwave_pb::hummock::compact_task::{PbTaskStatus, PbTaskType, TaskStatus, TaskType};
 use risingwave_pb::hummock::group_delta::DeltaType;
-use risingwave_pb::hummock::hummock_version::Levels as PbLevels;
-use risingwave_pb::hummock::hummock_version_delta::{ChangeLogDelta, GroupDeltas as PbGroupDeltas};
+use risingwave_pb::hummock::hummock_version::PbLevels;
+use risingwave_pb::hummock::hummock_version_delta::{ChangeLogDelta, PbGroupDeltas};
+use risingwave_pb::hummock::subscribe_compaction_event_request::PbReportTask;
 use risingwave_pb::hummock::{
-    HummockVersionDelta as PbHummockVersionDelta, PbHummockVersion, SstableInfo,
+    BloomFilterType, LevelType, PbCompactTask, PbHummockVersion, PbHummockVersionDelta,
+    PbInputLevel, PbKeyRange, PbLevel, PbLevelType, PbOverlappingLevel, PbSstableInfo,
+    PbTableStats, PbValidationTask, TableOption, TableSchema,
 };
+use serde::Serialize;
 
 use crate::change_log::TableChangeLog;
+use crate::key_range::KeyRange;
 use crate::table_watermark::TableWatermarks;
-use crate::{CompactionGroupId, HummockSstableObjectId};
+use crate::{CompactionGroupId, HummockSstableObjectId, ProtoSerializeSizeEstimatedExt};
+
+#[derive(Debug, Clone, PartialEq, Default, Serialize)]
+pub struct OverlappingLevel {
+    pub sub_levels: Vec<Level>,
+    pub total_file_size: u64,
+    pub uncompressed_file_size: u64,
+}
+
+impl From<&PbOverlappingLevel> for OverlappingLevel {
```

Comment:
I think it's unnecessary to implement a trait for the conversion, because we don't have any code that depends on the `From` or `Into` trait. The only benefit is that when we implement `From`, we can call `.into()`, but the benefit is limited. The code will be the same if we don't implement the trait and just implement a method for the new struct like
```
impl OverlappingLevel { // No impl `From<&PbOverlappingLevel>`
    fn from(pb_overlapping_level: &PbOverlappingLevel) -> Self {
        Self {
            ...
        }
    }
}
```

And then we'd better rename the method name from `from`, which is too general, to `from_protobuf` and `to_protobuf` so that we can explicitly know that we are doing a conversion between new struct and pb when we read the code.

ditto for all other `impl From` in this PR.

Reply:
In some cases, I want to consume the incoming arguments directly (instead of copying them), and if we want to fulfil this requirement, we still need to implement four functions
1. from_protobuf(struct)
2. from_protobuf(struct_ref)
3. to_protobuf1(struct)
4. to_protobuf2(struct_ref)

I don't think there's much difference except in naming, as in commit, and I've provided `from_protobuf` and `to_protobuf` wrappers for some common structures